### DISCUSSION TITLE
Egg Not Responding

### DISCUSSION THREAD

**Discussion by jamesonuk** [visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA#444] [Q&A]
I have just got around to trying out my device on the egg and I seem to be seeing very similar issues to https://github.com/visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA/discussions/431

I have created a separate discussion as I have built mine with a D1 mini and no PCB and also I have a slightly different model.
I have a 2021 Zurich with a 100102 pump.

Running the hardware test with the JST connectors plugged together says it is fine but when I plug it in I get the music, the display scrolls the IP address and date and then nothing.   Buttons on the egg and device do nothing.    The device is running and I can connect to the web interface and I checked there is about 5.2V on the power lines (when connected to egg only)

Following on from the comments in the other thread I connected the device up to the laptop via USB and connected up the serial monitor.   The buttons on the egg then sprang into life and changes on the egg were reflected in the UI, it picked up temperature and C/F state etc.   Changes in the UI however were not triggered on the pump..

I do have it hooked up to MQTT broker which is sending messages
```
{
    "CONTENT": "STATES",
    "TIME": 1675946868,
    "LCK": 0,
    "PWR": 0,
    "UNT": 1,
    "AIR": 0,
    "GRN": 0,
    "RED": 0,
    "HTR": 0,
    "FLT": 0,
    "CH1": 32,
    "CH2": 51,
    "CH3": 53,
    "HJT": 0,
    "BRT": 7,
    "TGT": 36,
    "TMP": 35,
    "VTF": 32.99998474,
    "VTMC": 32.94703674,
    "VTMF": 91.30466461,
    "AMBC": 20,
    "AMBF": 68,
    "AMB": 20,
    "VTM": 32.94703674,
    "TGTC": 36,
    "TMPC": 35,
    "TGTF": 97,
    "TMPF": 95
}
```

Looking for a bit of help to work out what isn't working or what I can try.

**Edit:** Also noticed https://github.com/visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA/issues/437 which seems to have been resolved by using the updated PCB layout but this shouldn't matter if just going direct to the D1 Mini and the hardware test says that is wired correctly? (Programming I know about, electronics not so much 😜)

**Comment by visualapproach** [comment]
Check correlation between pressed button on the display and what's reported over mqtt or websockets

**Reply by jamesonuk** [reply]
I didn't actually consider looking at the MQTT side of things and assumed if the web interface didn't work then the MQTT one wouldn't either.

I think the physical buttons not responding was me being impatient.   It seems to take about 60 seconds for for the power button to come on and after that the physical buttons work.

Seems that once the power button does light up I can control the tub via Home Assistant quiet nicely (but the web UI is still not controlling the egg).   I will be controlling it via HA anyway so it isn't a massive issue.

Is odd though why the web UI isn't working but MQTT is (and why it seems to come to life when you connect the serial monitor).

Going to do some debugging (particularly around powering on / off) to see if I can work out what is going on.

@visualapproach have sent you a coffee (well beer) 🍻 thanks for creating this

**Reply by visualapproach** [reply]
@jamesonuk you ROCK, thank you 😃

**Comment by jamesonuk** [comment]
Going back to this as woke up today to an E02 error so duly went out and changed the filter but when I reset the pump I got the same issue again.   I get the jingle, the IP address and date scroll on the display but then nothing.   The buttons on the egg, MQTT nor web UI are doing anything.

The Web UI looks like the board can't process anything as I get 
![image](https://user-images.githubusercontent.com/1040621/223427858-2c9d9cb7-f70e-4ccb-b255-d21fb079c3e9.png)
running connected to the control side of the egg but not the display I get [***] so it is recognising that something has changed....

It has been running fine for about a month so I know the hardware / wiring is OK.   The egg works fine without the board.

Seems it needs something to kick it into life but I have no idea where to even start debugging this.   Any pointers as this has been great being able to control the tub via Home Assistant but a bit lost as to where to start.

**Comment by chunkysteveo** [comment]
What's your outdoor temperature? I know here in the UK we have a cold snap, and my tub was in error as the cold temp 'frost protect' thing tried to kick in... turned it all off.

**Reply by jamesonuk** [reply]
> What's your outdoor temperature? I know here in the UK we have a cold snap, and my tub was in error as the cold temp 'frost protect' thing tried to kick in... turned it all off.

Yeah in the midlands so was a bit chilly last night but it is definitely the dirty filter that caused the E02 and replacing the filter has it all working nicely without the controller attached.   The issue is with the controller just not seeming to correctly link the egg controller and the display.   I had the same when I first set it up but then seemed to spring into life and worked until I reset the power.

Had a flurry of snow earlier but seems to be warming up now so will try and pop out and try again later.   It does seem like https://github.com/visualapproach/WiFi-remote-for-Bestway-Lay-Z-SPA/discussions/431#discussioncomment-4757070 where there is something that just kicks things into life

**Reply by visualapproach** [reply]
> What's your outdoor temperature? I know here in the UK we have a cold snap, and my tub was in error as the cold temp 'frost protect' thing tried to kick in... turned it all off.

Seems like it may be the case. Judging from the screenshot the CIO is not sending anything to the ESP, or if it does it doesn't arrive at the ESP. It could also be corroded contacts etc

**Reply by jamesonuk** [reply]
> > What's your outdoor temperature? I know here in the UK we have a cold snap, and my tub was in error as the cold temp 'frost protect' thing tried to kick in... turned it all off.
> 
> Seems like it may be the case. Judging from the screenshot the CIO is not sending anything to the ESP, or if it does it doesn't arrive at the ESP. It could also be corroded contacts etc

water temp was still over 30C so frost thing would not have kicked in as its based on water not ambient temperature.

Is there anyway I can see what if anything is being received?

**Comment by franpalacio** [comment]
Hello, regarding case #431, after trying all the possible combinations and in the absence of further testing, I just tried one more time and it works!!! Although I had already tried this configuration and it didn't work, now miraculously it has started to work. The only change that I have made, and that it occurs to me that it could be, has been to change the usb cable that I have used for programming. The cable that I used previously in theory worked correctly and I have programmed many other ESPs. So I don't understand it, but it's the only change I've made.
The configuration that I have used has been:
MIAMI2021
PCB_V2B
Try programming it with another higher quality cable.
All the best

**Comment by jamesonuk** [comment]
OK had a little time so I did a couple of things.   First one was to build a new controller.   The first one was a bit hacked together but I have built something a little better and cleaner (still just using the d1 mini).

So plugged it in and the same.   Display plays the tune and shows IP / date but, there appears to be a light behind the C/F part of the display but no buttons work and the web UI shows the disconnected data as above.

I got the laptop out and hooked up via USB and serial monitor but still nothing.

Then I just did various things to plug and unplug the controller (connecting with power to pump totally off, connecting with power on, controller first, display first....)

At some point it came to life. the display started working and the web UI picked up the right info.

My hunch at the moment is that is seems the D1 wants to be running before it is connected to egg (hence why the USB power connection seems to work).   It is almost as if there is a handshake between the controller and display that is being missed why the D1 boots.

Not sure if this is related to using the D1, my pump model (S100102) or something I am doing wrong but it is totally repeatable and there seem to be a few threads of people seeing similar things.

**Reply by visualapproach** [reply]
Okay, the 'handshake' would likely be: ordinary traffic must take place at start or the cio gives up. Should be easy to check by hooking a oscilloscope or even a LED to the one of the data lines from cio. I can see if there is a way to speed up the start sequence.

**Reply by jamesonuk** [reply]
> Okay, the 'handshake' would likely be: ordinary traffic must take place at start or the cio gives up. Should be easy to check by hooking a oscilloscope or even a LED to the one of the data lines from cio. I can see if there is a way to speed up the start sequence.

I don't have an oscilloscope and I an not sure what exactly is needed to trigger this system into life but has happened with multiple versions of the controller (albeit same hardware) so I don't think it is just my bad soldering 🤪

If the kids ever let me het some time I will attempt to some more scientific experiments with pulling the power and starting up again and seeing if I can figure out what the actual trigger to bring it to life is

**Reply by visualapproach** [reply]
I pushed a minor update that starts listening to the cio a few ms earlier. May or may not do any good.

**Reply by visualapproach** [reply]
What happens if you start the cio (pump) without the display connected, and after a while connect it? No esp involved. If it fails then it's probably some time limit before the cio panics. If it works then it's something else. Like a brownout due to lack of power on the 5V. Check the file rebootinfo.txt and see the reason for your reboots, if there were any.

**Reply by jamesonuk** [reply]
been a bit side tracked with other stuff but I noticed the tub temperature had dropped down and realised that the controller seemed to have  died.    I restarted everything and back to square one in that it wasn't firing up correctly so no MQTT and the web interface showing the default values (which look as if they are different in v4).

Anyway a little bit of playing around roughly the same as before (no scrolling IP / date which I guess is a change in V4?)   What I did notice though was that I was getting the sounds when pressing the temperature up/down, power and timer buttons but nothing on the lock, pump, heat or air buttons.    

Taking out the controller and plugging the display straight in (I left it 30 seconds or so not being connected) and it sprang into life straight away.   Putting the controller  back and back to unresponsive display.    Back to a direct connection and it worked fine.

Tried powering the controller via USB and then connecting as well as connecting and then powering via USB but nothing.
Attaching to my laptop and starting a serial monitor and then plugging in the controller and everything sprang into life.

When I get some time I will try and do some proper testing of different scenarios but wasn't sure whether the fact some buttons were playing a sound and others not might give you any idea of what might be happening?

**Reply by visualapproach** [reply]
The IP and FW date is printed to the display in v4 too. 
Some suggestions:
Update to latest dev (beta) version 2023-04-09-1840 (Yeah I haven't made it through github actions yet)
Make the hardware test (Since you don't have LLC it should be fine with USB power)
Check reboot.txt for unexpected reboot reasons.
Change to pinout according to PCB_V2B if you haven't already.
My guesses:
Bad or intermittent connections
Weak power
Bad luck (in general or specifically with the 3.3V only beeing enough to read as a high when the planets are right)

**Reply by jamesonuk** [reply]
> The IP and FW date is printed to the display in v4 too. 
Ah I can't say I have seen this but there again I haven't been particularly looking and it has been quite bright sunshine.

>Some suggestions: Update to latest dev (beta) version 2023-04-09-1840 (Yeah I haven't made it through github actions yet) Make the hardware test (Since you don't have LLC it should be fine with USB power) Check reboot.txt for unexpected reboot reasons. Change to pinout according to PCB_V2B if you haven't already. My guesses: Bad or intermittent connections Weak power Bad luck (in general or specifically with the 3.3V only beeing enough to read as a high when the planets are right)

The fact the display seems to work fine when connected directly must mean it is an issue with the controller rather than the egg but I dd try and rule some things out by building a new controller (albeit using the same model of MCU) but both controllers are exhibiting the same behaviour and with both v3 and v4.....     Might try with a different MCU (although I have a feeling the ones I have lying around are all ESP32s).

When I ever get a few hours I will try and do some proper debugging and also look at the release version stuff...

**Reply by visualapproach** [reply]
@jamesonuk beta version listens to layzspa.local/debug-on/ and .../debug-off/
While on it logs raw data from cio and dsp to "log.txt". You can run it a while and press some buttons.
